### PR TITLE
log-backup: use row-level memory usage statistic for initial scan

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/15714

What's Changed:
This PR removed the semaphore that may make TiKV panic, and refined the place where we consume the memory quota. Now, we will try to consume all messages as soon as we have reached the memory quota.

```commit-message
```
### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression

### Release note 

```release-note
Fixed a bug that may cause log backup's checkpoint get stuck with huge rows.
```

### CODE REVIEW

File: components/tikv_util/src/memory.rs [hunk 1]
```diff
@@ -87,6 +90,34 @@ pub struct MemoryQuota {
...
+
+impl OwnedAllocated {
+    pub fn of(target: Arc<MemoryQuota>) -> Self {
+        Self {
+            allocated: 0,
+            from: target,
+        }
+    }
+
+    pub fn allocate_more(&mut self, bytes: usize) -> Result<(), MemoryQuotaExceeded> {
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:106]:
```suggestion
    pub fn alloc(&mut self, bytes: usize) -> Result<(), MemoryQuotaExceeded> {
```

File: components/tikv_util/src/memory.rs [hunk 2]
```diff
@@ -87,6 +90,34 @@ pub struct MemoryQuota {
     capacity: AtomicUsize,
 }
 
+pub struct OwnedAllocated {
+    allocated: usize,
+    from: Arc<MemoryQuota>,
+}
+
+impl OwnedAllocated {
+    pub fn of(target: Arc<MemoryQuota>) -> Self {
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:99]: Why not `fn new`? It's idiomatic.

File: components/tikv_util/src/memory.rs [hunk 3]
```diff
@@ -87,6 +90,34 @@ pub struct MemoryQuota {
...
+            from: target,
+        }
+    }
+
+    pub fn allocate_more(&mut self, bytes: usize) -> Result<(), MemoryQuotaExceeded> {
+        let res = self.from.alloc(bytes);
+        if res.is_ok() {
+            self.allocated += bytes;
+        }
+        res
```
Comment (overvenus) [inline components/tikv_util/src/memory.rs:111]:
```suggestion
        self.from.alloc(bytes)?;
        self.allocated += bytes;
```

File: components/backup-stream/tests/suite.rs [hunk 4]
```diff
@@ -557,45 +582,72 @@ impl Suite {
...
+            let pbuf: &[u8] = &buf[range.0..range.1];
+            run_async_test(async {
+                decoder.write_all(pbuf).await.unwrap();
+                decoder.flush().await.unwrap();
+                decoder.close().await.unwrap();
+            });
+            decoder.into_inner()
+        };
+        for entry in files.write_cf {
+            println!("checking write: {:?}", entry);
```
Comment (overvenus) [inline components/backup-stream/tests/suite.rs:598]:
```suggestion
            debug!("checking write: {:?}", entry);
```

**Review by overvenus — APPROVED** [review]

File: components/backup-stream/src/event_loader.rs [hunk 5]
```diff
@@ -384,40 +389,44 @@ where
...
-            let no_progress = event_loader.entry_batch.is_empty();
-            let stat = stat?;
+            let mut allocated = OwnedAllocated::new(Arc::clone(&self.quota));
+            let (res, disk_read) =
+                utils::with_record_read_throughput(|| event_loader.fill_entries(&mut allocated));
+            let res = res?;
             self.with_resolver(region, handle, |r| {
                 event_loader.emit_entries_to(&mut events, r)
             })?;
-            if no_progress {
```
Comment (3pointer) [inline components/backup-stream/src/event_loader.rs:400]: so still exit too early before :)

**Review by 3pointer — APPROVED** [review]
LGTM

File: components/backup-stream/src/event_loader.rs (hunk 5 above)
Comment (YuJuncen) [reply components/backup-stream/src/event_loader.rs:400]: In fact now we are exit earlier than before. In the past we exit once we have noticed the scan *has returned* nothing, for now the scan will report whether it *will return* nothing in the previous run of scanning.

**Pull request merged**